pytestqt/plugin.py

```diff
@@ -421,6 +426,98 @@ def pytest_configure(config):
         "qt_no_exception_capture: Disables pytest-qt's automatic exception "
         'capture for just one test item.')
 
+    if config.getoption('qt_log'):
+        config.pluginmanager.register(QtLoggingPlugin(config), '_qt_logging')
+
 
 def pytest_report_header():
-    return ['qt-api: %s' % QT_API]
\ No newline at end of file
+    return ['qt-api: %s' % QT_API]
+
+
+class QtLoggingPlugin(object):
+    """
+    Pluging responsible for installing a QtMessageHandler before each
+    test and augment reporting if the test failed with the messages captured.
+    """
+
+    def __init__(self, config):
+        self.config = config
+
+    def pytest_runtest_setup(self, item):
+        item.qt_log_capture = _QtMessageCapture()
+        previous_handler = qInstallMsgHandler(item.qt_log_capture._handle)
+        item.qt_previous_handler = previous_handler
+
+    @pytest.mark.hookwrapper
+    def pytest_runtest_makereport(self, item, call):
+        """Add captured Qt messages to test item report if the call failed."""
+
+        outcome = yield
+        report = outcome.result
+
+        if call.when == 'call':
+
+            if not report.passed:
+                long_repr = getattr(report, 'longrepr', None)
+                if hasattr(long_repr, 'addsection'):
+                    lines = []
+                    for msg_type, msg in item.qt_log_capture.messages:
+                        msg_name = _QtMessageCapture.get_msg_name(msg_type)
+                        lines.append('{0}: {1}'.format(msg_name, msg))
+                    if lines:
+                        long_repr.addsection('Captured Qt messages',
+                                             '\n'.join(lines))
+            # Release the handler resources.
+            qInstallMsgHandler(item.qt_previous_handler)
+            del item.qt_previous_handler
+            del item.qt_log_capture
+
+
+class _QtMessageCapture(object):
+    """
+    Captures Qt messages when its `handle` method is installed using
+    qInstallMsgHandler, and stores them into `messages` attribute.
+
+    :attr messages: list of Message named-tuples.
+    """
+
+    Message = collections.namedtuple('Message', 'msg_type, msg')
+
+    def __init__(self):
+        self._messages = []
+
+    def _handle(self, msg_type, msg):
+        """
+        Method to be installed using qInstallMsgHandler, stores each message
+        into the `messages` attribute.
+        """
+        if isinstance(msg, bytes):
+            msg = msg.decode('utf-8', errors='replace')
+        self._messages.append(self.Message(msg_type, msg))
+
+    @property
+    def messages(self):
+        """Acess messages captured so far.
```

Comment:
Hehe, I thought so :stuck_out_tongue: 

There are some new interesting concepts/things here for me to learn, so I thought I'll take a quick look whenever I feel like and have time, and then take a more proper look whenever you say it's done :wink: